### PR TITLE
Solved "Counting bits"

### PR DESCRIPTION
**Pull request by makcgr**
Counting Bits:
- solved in naive way and also looked up and implemented the "pro way" with binary arithmetics.
- "Pro" way seems to be more laconic, but not exceptionally faster (in fact, it's about the same speed).

Also did "Climbing Stairs" closing my "debt gap", eventually I looked it up.
Check out the second algorithm: a very cool way of resolving the count of steps without Fibonacci in one pass!

### CODE REVIEW

**Comment by rokovalenko** [comment]
"pro" solution looks awesome. I can't even grasp the mindset behind bitwise operations. There are so many things to learn

File: mgrachev/counting bits/Program.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+        int[] arr = new int[bitDepth];
+        Array.Fill(arr, 0);
+
+        int curr = number;
+        int bitsCount = 0;
+
+        while (curr != 0)
+        {   
+            var log2 = (int) Math.Log2(curr);
```
Comment (PShchegolevatykh) [inline mgrachev/counting bits/Program.cs:55]: Interesting take on this.

File: mgrachev/counting bits/Program.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        return bitsCount;
+    }
+
+    /*
+    Pro solution: O(N*logN) time complexity (all numbers * time to iterate all bits)
+    Will use bitwise operators to do the shift to the right and check the least significant bit (LSB).
+    */
+
+    public static int CountBitsInNumberPro(int number)
```
Comment (PShchegolevatykh) [inline mgrachev/counting bits/Program.cs:70]: Incredible! I was thinking about bitwise operations including shifts but did not get to the solution. LSB for the win!

File: mgrachev/number of ways to climb stairs/Program.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        return Fib(n-1) + Fib(n-2);
+    }
+
+
+    /*
+     Solution 2
+     Here we do computations iteratively
+     from the bottom up.
+     Let's visualize the step ladder e.g. for n=5 with 5 steps:
```
Comment (PShchegolevatykh) [inline mgrachev/number of ways to climb stairs/Program.cs:41]: Great comments as always, even ASCII art! :)

File: mgrachev/number of ways to climb stairs/Program.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    NOTE: there is extra (N+1)th cell to store the overall number of ways.
+
+    For first two steps, n of ways to climb there is 1.
+    Now for third step, we can climb there with sum of ways for two previous steps.
+    We continue like this for the rest of ladder (array) until we reach the top (array[n]):
+    ways(step[N]) = ways(step[N-1]) + ways(step[N-2])
+
+    Time complexity: O(N), since this is a single pass.
+    */
+    public int ClimbStairs2(int n)
```
Comment (PShchegolevatykh) [inline mgrachev/number of ways to climb stairs/Program.cs:64]: Mind-blowing solution! 🤯

File: mgrachev/number of ways to climb stairs/Program.cs (hunk 3 above)
Comment (makcgr) [reply mgrachev/number of ways to climb stairs/Program.cs:41]: Thanks! Yeah, ASCII art was really handy here)

File: mgrachev/counting bits/Program.cs (hunk 2 above)
Comment (makcgr) [reply mgrachev/counting bits/Program.cs:70]: Yeah, it's cool that you already thought about it! Sometimes it's hard to think outside of the box) Well, now we know :)

**Comment by makcgr** [comment]
> 



> "pro" solution looks awesome. I can't even grasp the mindset behind bitwise operations. There are so many things to learn

Yep, me too, I did not think about the bitwise ops until I looked it up!

**Pull request merged**